WcaOnRails/lib/competition_results_validator.rb

```diff
@@ -0,0 +1,606 @@
+
+        # Checks for time limits if it can be user-specified
+        if !["333mbf", "333fm"].include?(result.eventId)
+          cumulative_wcif_round_ids = time_limit_for_round.cumulative_round_ids
+          # Now let's try to find a DNF/DNS result followed by a non-DNF/DNS result
+          # Do the same for DNS.
+          has_result_after = { SolveTime::DNF => false, SolveTime::DNS => false }
+          has_result_after.keys.each do |not_complete|
+            first_index = all_solve_times.find_index(not_complete)
+            if first_index
+              # Just use '5' here to get all of them
+              solves_after = all_solve_times[first_index, 5]
+              has_result_after[not_complete] = solves_after.select(&:complete?).any?
+            end
+          end
+
+          # Always output the warning about DNS followed by result
+          if has_result_after[SolveTime::DNS]
+            @warnings[:results] << format(DNS_AFTER_RESULT_WARNING, round_id: round_id, person_name: person_info.name)
+          end
+
+          case cumulative_wcif_round_ids.length
+          when 0
+            # easy case: each completed result (not DNS, DNF, or SKIPPED) must be below the time limit.
+            results_over_time_limit = completed_solves.select { |t| t.time_centiseconds > time_limit_for_round.centiseconds }
+            if results_over_time_limit&.any?
+              @errors[:results] << format(RESULT_OVER_TIME_LIMIT_ERROR, round_id: round_id, person_name: person_info.name, time_limit: time_limit_for_round.to_s(round_info))
+            end
+          else
+            # Handle both cumulative for a single round or multiple round by doing the following:
+            #  - gather all solve times for all the rounds (necessitate to map round's WCIF id to "our" round ids)
+            #  - check the sum is below the limit
+            #  - check for any suspicious DNF result
+
+            # Match wcif round ids to "our" ids
+            cumulative_round_ids = cumulative_wcif_round_ids.map do |wcif_id|
+              parsed_wcif_id = Round.parse_wcif_id(wcif_id)
+              # Get the actual round_id from our expected rounds by id
+              @expected_rounds_by_ids.select do |id, round|
+                round.event.id == parsed_wcif_id[:event_id] && round.number == parsed_wcif_id[:round_number]
+              end.first[0]
+            end
+
+            # Get all solve times for all cumulative rounds for the current person
+            all_results_for_cumulative_rounds = cumulative_round_ids.map do |id|
+              results_by_round_id[id].find { |r| r.personId == result.personId }
+            end.compact.map(&:solve_times).flatten
+            completed_solves_for_rounds = all_results_for_cumulative_rounds.select(&:complete?)
+            number_of_dnf_solves = all_results_for_cumulative_rounds.select(&:dnf?).size
+            sum_of_times_for_rounds = completed_solves_for_rounds.map(&:time_centiseconds).reduce(&:+) || 0
```

Comment:
Nit: `.map(&:time_centiseconds).reduce(&:+)` -> `.sum(&:time_centiseconds)`. Also `sum` default to `0` if the array is empty =)